src/wallet/wallet.cpp

```diff
@@ -6896,10 +6898,28 @@ PaymentAddressSource GetSourceForPaymentAddress::operator()(const libzcash::Sapl
         // also have the corresponding SaplingExtendedFullViewingKey.
         if (m_wallet->GetSaplingIncomingViewingKey(zaddr, ivk)) {
             if (m_wallet->HaveSaplingFullViewingKey(ivk)) {
-                // If we have the HD keypath, it's related to the legacy seed
+                // If we have the HD keypath, it's related to a seed.
                 if (m_wallet->mapSaplingZKeyMetadata.count(ivk) > 0 &&
                         m_wallet->mapSaplingZKeyMetadata[ivk].hdKeypath != "") {
-                    return PaymentAddressSource::LegacyHDSeed;
+                    // The following keypaths have been used in zcashd:
+                    // - Legacy seed:     m/32'/coin_type'/account_counter'
+                    // - Mnemonic phrase: m/32'/coin_type'/ZCASH_LEGACY_ACCOUNT'/account_counter'
+                    //
+                    // Thus if the keypath uses account ZCASH_LEGACY_ACCOUNT and
+                    // is longer than "m/32'/coin_type'/ZCASH_LEGACY_ACCOUNT'",
+                    // it is derived from the mnemonic.
+                    auto account = libzcash::ParseHDKeypathAccount(
```

Comment:
The way that I've handled this determination is to check against the seed fingerprint for the mnemonic seed; if the fingerprint in `mapSaplingZKeyMetadata` doesn't match the fingerprint of the mnemonic seed, we treat it as a legacy address.